lib/tree.rb

```diff
@@ -2,54 +2,130 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+  
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  def add_helper(current_node, new_node)
+    return new_node if current_node.nil?
+
+    if new_node.key < current_node.key
+      current_node.left = add_helper(current_node.left, new_node)
+    elsif new_node.key > current_node.key
+      current_node.right = add_helper(current_node.right, new_node)
+    end 
+
+    return current_node
+  end 
+
+  # Time Complexity: o(log(n))
+  # Space Complexity: o(1)
+  def add(key, value = nil)
+    new_node = TreeNode.new(key, value)
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current_node,key)
+    return nil if current_node.nil? 
+
+    if current_node.key == key
+      return current_node.value
+    elsif key <= current_node.key
+      find_helper(current_node.left, key)
+    else key > current_node.key
+      return find_helper(current_node.right, key)
+    end 
+  end 
+
+
+  # Time Complexity: o(n)
+  # Space Complexity: o(n)
   def find(key)
```

Comment:
👍 the time complexity is O(log n) if the tree is balanced and O(n) if it's unbalanced.  The space complexity is the **same** due to recursion.